### PR TITLE
Add chat input sanitization

### LINKED ISSUE #47217: VL models fail when their image placeholder is in user text.
### System Info

- `transformers` version: 5.13.0
- Platform: macOS-26.3.1-arm64-arm-64bit-Mach-O
- Python version: 3.14.3
- Huggingface_hub version: 1.23.0
- Safetensors version: 0.8.0
- Accelerate version: not installed
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): not installed (NA)
- Using distributed or parallel set-up in script?: No

### Who can help?

@zucchini-nlp is the listed multimodal maintainer. 

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
# /// script
# requires-python = ">=3.14"
# dependencies = [
#     "torchvision>=0.28.0",
#     "transformers>=5.13.0",
# ]
# ///
import transformers


def main() -> None:
    processor = transformers.AutoProcessor.from_pretrained("LiquidAI/LFM2.5-VL-1.6B")

    messages = [
        {
            "role": "user",
            "content": [
                {"type": "text", "text": "How do I convert this to an <image> tag?"},
                {
                    "type": "image",
                    "url": "https://cdn-uploads.huggingface.co/production/uploads/61b8e2ba285851687028d395/2b08LKpev0DNEk6DlnWkY.png",
                },
            ],
        }
    ]

    print(processor.apply_chat_template(messages))

    # Throws ValueError: The number of images in the text [2] and images [1] should be the same.
    print(processor.apply_chat_template(messages, tokenize=True))


if __name__ == "__main__":
    main()
```

Run with `uv run repro.py`.

### Expected behavior

No failure, and correct tokenization.

### PR DESCRIPTION
**Pull request by Rocketknight1**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47386) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47386)

This PR adds special token sanitization for chat inputs. I debated trying to break up the special tokens by retokenizing them with normal tokens, or escaping them, but I think the simplest solution is just to remove them entirely, similar to `tokenizer.decode(skip_special_tokens=True)`. It should be quite uncommon for chat inputs to contain model special tokens, and when it does it's probably malicious, so I think it's fine if input is slightly distorted in those cases.

I also got a code agent to add some tests to make sure we catch various sneaky ways to smuggle in special tokens despite this!

Fixes #47217

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
this is fine for now, but in terms of perfs it won't really be optimal. I am happy to hvae this as a workaround. but I think we need to think a bit to come up with a better solution!

**Comment by bbrowning** [comment]
> It should be quite uncommon for chat inputs to contain model special tokens, and when it does it's probably malicious, so I think it's fine if input is slightly distorted in those cases.

There are legitimate cases, even if not extremely common. I use open weight models with vLLM and also work on vLLM, and I regularly run into these kinds of issues where models get confused because I'm editing a test file that has `<think>` type text in it that also happens to correspond to the model's actual thinking token delimiter.

And for inference server usage (as opposed to users directly applying chat templates), there are other considerations such as multimodal token placeholders that may prevent them from being able to use the `sanitize_special_tokens` approach described here because real special tokens may get inserted into user input and need to survive chat template rendering. If we take an example from a Qwen2-Audio model, we have placeholders like `Audio 1: <|audio_bos|><|AUDIO|><|audio_eos|>` - `<|audio_bos|>` is an example of a `special` token that would get sanitized away. We use these kinds of placeholders less on newer models, and this may be something that becomes obsolete once we stop supporting models that don't natively support multimodal content parts in their chat templates.

---

For completeness' sake, here are two other options that some of us are prototyping for this problem on the inference server side:

1. Replacing untrusted content (user inputs, tool outputs, etc) with sentinels and rendering the chat template with sentinels then using that to tokenize trusted vs untrusted content separately, where the untrusted tokenizer cannot produce the special / structural token ids the model uses for turn boundaries, reasoning content, etc.
2. Replacing all structural text in a chat template with randomized placeholders that we strip from all user content and tokenizing this with a new tokenizer that maps those random placeholders to the actual token ids the model needs to see. The net effect is the user would be unable to pass in text that tokenizes to the special tokens without risk of it rejecting legitimate content.

What both of these approaches do is ensure the user content - whether legitimate or a prompt injection attempt - cannot tokenize to the actual structural / special token ids. Attempts to do so will just result in the text tokenizing to subword tokens instead of the special tokens, meaning they will decode to the same text equivalent but the model should not treat them as it would the actual special tokens it has been trained on. One thing to note is that not all important structural tokens in models are flagged as "special", so this can't just use the pre-existing `split_special_tokens` to properly sanitize or demote to subwords.

---

None of this is meant to say adding a `sanitize_special_tokens` would be bad. This is a real problem we need to fix to properly turn a list of chat messages into token ids - both for correctness' sake and security. However, this also won't be sufficient for most inference servers, and is a general problem we need to solve in this space that may involve tokenizing untrusted content and trusted content differently as well as a more nuanced, model-specific understanding of what tokens are important structural markers in conversations and which are not.

cc @yzong-rh who has been helping me prototype and think about some of these things.

**Comment by samuelstevens** [comment]
From the initial PR comment by @Rocketknight1:

> It should be quite uncommon for chat inputs to contain model special tokens, and when it does it's probably malicious, so I think it's fine if input is slightly distorted in those cases.

I agree with the @bbrowning's comment on this not being true. LFM2-VL models use `<image>` as their placeholder token, which is common enough in HTML/XML/SVG contexts, and editing SVGs is a common goal for models that can handle image inputs.

And from @bbrowning's comments

> This is a real problem we need to fix to properly turn a list of chat messages into token ids

Agree. Ideally, we would never turn our structured representation of chat messages into a flat string via `apply_chat_template` before converting to token ids, and instead just go straight from chat messages to token ids.  It seems that other neolabs are figuring this out as well. See my comment in https://github.com/huggingface/transformers/issues/47217

> It seems that both Prime Intellect and Thinking Machines developed solutions to these problems ([primeintellect.ai/blog/renderers](https://www.primeintellect.ai/blog/renderers) and [tinker-docs.thinkingmachines.ai/tutorials/core-concepts/rendering](https://tinker-docs.thinkingmachines.ai/tutorials/core-concepts/rendering/)). Curious if HF is aware of these issues and has any plans?

From an implementation perspective:

> Replacing all structural text in a chat template with randomized placeholders that we strip from all user content and tokenizing this with a new tokenizer that maps those random placeholders to the actual token ids the model needs to see. 

I've had this suggested to me as well, and while it does solve the problem, I think it makes it very hard to debug, because there are random values that will inherently be different at every step.

**Comment by Rocketknight1** [comment]
After thinking about it I agree with the commenters that simply dropping special tokens is going to be breaking. I'm experimenting with the random-placeholder approach, we'll see how that looks!

File: src/transformers/utils/chat_template_utils.py [hunk 1]
```diff
@@ -629,3 +630,302 @@ def __init__(self, messages: dict):
             if not is_valid_message(message):
                 raise ValueError("When passing chat dicts as input, each dict must have a 'role' and 'content' key.")
         self.messages = messages
+
+
+@lru_cache
```
Comment (ArthurZucker) [inline src/transformers/utils/chat_template_utils.py:635]: https://github.com/daac-tools/daachorse

if you want something fast, it has python bindings.

File: src/transformers/utils/chat_template_utils.py [hunk 2]
```diff
@@ -629,3 +630,302 @@ def __init__(self, messages: dict):
...
+    substitutions = {}
+    pattern = _compile_special_token_pattern(tuple(sanitization_special_tokens(tokenizer)))
+    conversations = _sanitize_chat_input(conversations, pattern, substitutions)
+    tools = _sanitize_chat_input(tools, pattern, substitutions)
+    documents = _sanitize_chat_input(documents, pattern, substitutions)
+    kwargs = _sanitize_chat_input(kwargs, pattern, substitutions)
+    return conversations, tools, documents, kwargs, substitutions
+
+
+def _sanitize_chat_input(chat_input: Any, pattern: re.Pattern | None, substitutions: dict[str, str]) -> Any:
```
Comment (ArthurZucker) [inline src/transformers/utils/chat_template_utils.py:664]: I really don't think recursion is something we want here no?

**Review by ArthurZucker — CHANGES_REQUESTED** [review]
Text-level escaping is fundamentally a fragile text-level hack.
We should completely decouple template structure from untrusted content generation. The `split_special_tokens=True` is the key. 

Basically break down jinja to let it do the "format" part, while the tokenizer does the encoding, alone. 

```python
# A cleaner, token-safe isolation layer
class UntrustedInput:
    def __init__(self, content: str, tokenizer):
        self.content = content
        self.tokenizer = tokenizer

    def tokenize_securely(self):
        # Let the tokenizer do what it was designed to do:
        # Treat the input strictly as literal string primitives.
        return self.tokenizer.encode(self.content, add_special_tokens=False)
```
And in the Jinja macro:
```jinja
<|im_start|>user
{{ message.content.tokenize_securely() }}<|im_end|>
```

This ensures:
- Native Performance: We trust the low-level rust/C++ tokenizer backend constraints rather than maintaining fragile Python regex loops.
- Separation of Concerns: Jinja dictates formatting topology; the tokenizer enforces safety boundaries.

This is also much much simpler.

**Comment by Rocketknight1** [comment]
@ArthurZucker I'm not sure that solution makes sense! Jinja operates on strings, not tokens, so we can't do the tokenization until **after** Jinja formats everything. It would be nice, but we can't cleanly tokenize user strings first and just input that to Jinja.

File: src/transformers/utils/chat_template_utils.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/utils/chat_template_utils.py:635]: I checked - Aho-Corasick is a nice clean way to scan for multiple strings at once, but this scan is very short compared to tokenization, so we probably don't need the external dependency when `re` can do it!

**Comment by ArthurZucker** [comment]
@Rocketknight1 please have a look at the absolute vibe coded PR. I am pretty sure we can do even less LOC. Let's aim for simplicity and try to understand the tools we are using.

**Comment by yunusdim** [comment]
Small thing to add on @bbrowning's point that not every structural token is flagged `special`. That's the gap I'd make sure a test actually pins down, since it survives whichever approach wins here.

Whatever ships, escaping or the token-level path in #48052, it's easy to test the obvious case (`<|im_start|>` in user content, check it's neutralized) and miss the real one. The case that bit me: models where the role markers are registered as added/special tokens but the reasoning or section delimiters aren't. Same class of marker, but `split_special_tokens` never sees the second kind, because it was never special to begin with.

So the assertion I'd want green: take a model like that, drop the non-special structural strings into user content, run `apply_chat_template`, and check they come out as subword ids, not the control id the template's own markers produce. If a content-origin marker lands on the same id as a real one, there's still a hole.

Not tied to this PR's shape, just the test I'd want before calling it closed. I wrote the general rule up (content shouldn't be able to resolve to the container's own control ids) with a small behavioral bank, in case it's useful: https://doi.org/10.5281/zenodo.22722524

**Comment by Rocketknight1** [comment]
@arthurzucker you were totally right, a much simpler solution was possible! It took some iteration but I've got something much simpler that still works. I dropped support for the `assistant_tokens_mask` because this is usually used in training, but sanitizing special tokens is usually done for inference. Without those edge cases, the whole thing is 65 lines plus a test

**Comment by ArthurZucker** [comment]
On it!
